### PR TITLE
[WIP] No longer allow duplicate account names

### PR DESCRIPTION
**Pull request by CharlesEkkel**
## Description

Issue #316 has demonstrated that duplicate account names are currently permitted - the existing behaviour is that following the creation of duplicate accounts, there will henceforth be two entries in any autocompletion dropdown with the same name. They are logically distinct, but difficult to differentiate.

Please note that given this is my first PR for this repo, I'd very much appreciate feedback on how to improve it! I figured I'd start with something small.

## Changes

- Added a new validation rule which checks if the incoming account name matches any of the existing accounts.
- Added a function in `packages/loot-core/src/shared/accounts.js` which returns different validation error messages based on an input `errors.name`. I've done this to move the conditional logic out of the component and make it easier to add more validation rules in future if necessary.
- The default error message is the same as the one for `'required'` because this is the current behaviour (currently the value of the error name is not checked at all). I'm happy to add a different default message to be more comprehensive if necessary.

## Demo and Screenshots

See a demo running this code here: https://actual-demo.fly.dev

2022-10-29T12:34:02,800958803+11:00

### CODE REVIEW

**Comment by CharlesEkkel** [comment]
I've noticed that before and after videos are quite popular here, so here we go :)

### Before

https://user-images.githubusercontent.com/54133784/198814261-32671226-ce75-481c-ba8a-53f7799b30da.mp4

### After

https://user-images.githubusercontent.com/54133784/198814292-c617d237-b5fa-470e-97da-76a90aba6db6.mp4

**Review by MatissJanis — COMMENTED** [review]
I'm not strictly against this solution, but does it fully solve the problem? People will still be able to edit account names and thus end up in the same situation.

Have you considered doing the validation server-side? We could reject creation and edit requests if account names clash. And then show a nice message in the UI.

Apart from this conceptual design question: current code looks good :+1:

**Comment by rich-howell** [comment]
> I'm not strictly against this solution, but does it fully solve the problem? People will still be able to edit account names and thus end up in the same situation.
> 
> Have you considered doing the validation server-side? We could reject creation and edit requests if account names clash. And then show a nice message in the UI.
> 
> Apart from this conceptual design question: current code looks good 👍

Given the above, I have removed the waiting-review tag and moved it back to in progress so it can be revewied.

**Comment by rich-howell** [comment]
#316

**Comment by rich-howell** [comment]
I have reviewed this and can confirm that the validation rule works as expected when creating the account however as @MatissJanis pointed out it is still possible to edit the account name and be back where you started.

Is this something you are still willing to work on @CharlesEkkel?

**Comment by CharlesEkkel** [comment]
@rich-howell Thanks for the reminder, I'll get back to it when I have some time this week - Matiss made an excellent point. Also just saw the new release, very excited to try it out :smile:

**Comment by CharlesEkkel** [comment]
Will look into the failing tests - sorry for leaving this small PR dead in the water for so long. Things haven't been great lately.

But regardless, I'd greatly appreciate feedback on this approach - I've tried to find everywhere that tries to create or update the name of an account and added logic to check against existing names beforehand. However, I haven't been able to determine the preferred approach to signalling bad input behaviour like this...would throwing & catching errors be preferable to the error strings I've been returning & checking instead? 
The advantage to this approach is that ideally, once typescript is eventually in use, the compiler would automatically encourage you to handle the error cases. But I appreciate that throwing errors can indicate intent a lot better.

**Comment by j-f1** [comment]
Maybe there could be two different approaches?

1. When creating an account return an error if you try to create a duplicate name
2. If you try to rename an account to something duplicate (or empty!) then just throw the change away and go back to an old name

**Comment by CharlesEkkel** [comment]
Thanks @j-f1, excellent point about empty names, thats worth checking for as well.

And sorry, I don't think I was very clear - I was more talking about the preferred coding pattern for propogating errors so that they can be appropriately dealt with. E.g. throwing actual errors or returning strings representing different error states. 

But I think you're right about having two approaches - as the code is now, the original validation in the videos above still occurs, but if you try to update an account name to an existing account, then the change is rejected and an error message appears for a few moments. Does that sound reasonable?

**Comment by j-f1** [comment]
Sure, that makes sense! I like the notification part too. Actual doesn’t seem to use thrown errors much so sticking with an error return value seems like a good idea.

Also heads up that you have some failing tests over in CI.

**Comment by CharlesEkkel** [comment]
Thanks, I really appreciate the input. And yep 😢 I'll have a look later.

File: packages/loot-core/src/server/db/index.js [hunk 1]
```diff
@@ -565,14 +565,31 @@ export async function insertAccount(account) {
...
+
+  if (
+    accounts.some(
+      existingAccount =>
+        existingAccount.name === account.name &&
+        // If it's the current account name, then there's no issue.
+        existingAccount.id !== account.id
+    )
+  ) {
+    return 'Account name already exists';
```
Comment (j-f1) [inline packages/loot-core/src/server/db/index.js:590]: I’m not a fan of overloading the return value here. Can you refactor this function to return either `{ error: 'already-exists' }` or `{ id }`? That way the caller can check `if (error) return { error }` to propagate the error.

**Comment by MatissJanis** [comment]
👋 Hey @CharlesEkkel 

Just checking in: are you planing to pick this up again?

**Comment by CharlesEkkel** [comment]
@MatissJanis Thanks for the callout, probably eventually...but is there a way to mark this so that someone else can feel free to pick it up? Otherwise probably just close it, if no-one else picks up the issue I'll finish it & re-open a PR when things are a bit better. Sorry all.

**Comment by MatissJanis** [comment]
> @MatissJanis Thanks for the callout, probably eventually...but is there a way to mark this so that someone else can feel free to pick it up? Otherwise probably just close it, if no-one else picks up the issue I'll finish it & re-open a PR when things are a bit better. Sorry all.

Thanks for letting us know! We definitely do want this feature/fix, but take your time. Life should take priority over side-projects. Totally understand and support that :) 

I'll close this off for now, but please feel free to open a new PR if you resume working on this.

**Pull request closed without merging**